test/WebJobs.Script.Tests/Rpc/LanguageWorkerChannelTests.cs

```diff
@@ -56,6 +57,88 @@ public LanguageWorkerChannelTests()
                0);
         }
 
+        [Fact]
+        public void Dispose_WaitsForTimeout()
+        {
+            TestLogger logger = new TestLogger("LanguageWorkerChannelTests");
+            LanguageWorkerChannel channel = new LanguageWorkerChannel(
+               _workerId,
+               _scriptRootPath,
+               _eventManager,
+               _testWorkerConfig,
+               _mockLanguageWorkerProcess.Object,
+               logger,
+               _mockMetricsLogger.Object,
+               0);
+            channel.DisposeTimeoutHandle = new System.Threading.ManualResetEvent(true);
+            channel.SetupFunctionInvocationBuffers(GetTestFunctionsList("node"));
+            ScriptInvocationContext scriptInvocationContext = new ScriptInvocationContext()
+            {
+                FunctionMetadata = GetTestFunctionsList("node").FirstOrDefault(),
+                ExecutionContext = new ExecutionContext()
+                {
+                    InvocationId = Guid.NewGuid(),
+                    FunctionName = "js1",
+                    FunctionAppDirectory = _scriptRootPath,
+                    FunctionDirectory = _scriptRootPath
+                },
+                BindingData = new Dictionary<string, object>(),
+                Inputs = new List<(string name, DataType type, object val)>()
+            };
+            channel.SendInvocationRequest(scriptInvocationContext);
+            (channel as IDisposable)?.Dispose();
+            var traces = logger.GetLogMessages();
+            Assert.True(traces.Any(m => string.Equals(m.FormattedMessage, "DisposeTimeoutHandle has been set while waiting")));
+            Assert.True(traces.Any(m => string.Equals(m.FormattedMessage, "Waiting for buffer to drain for upto 5 seconds")));
+            Assert.True(traces.Any(m => string.Equals(m.FormattedMessage, "Done waiting for buffer drain")));
+        }
+
+        [Fact]
+        public async Task Dispose_WaitsForBufferDrain()
+        {
```

Comment:
Thanks for adding more tests. We need another variation where invocation responses are received simultinatenously over multiple threads. We need to ensure the check that sets drainBufferHandle is thread safe.